File: Difficulty/Medium/[59] Spiral Matrix II/solution.cpp
/*
 * @lc app=leetcode id=59 lang=cpp
 *
 * [59] Spiral Matrix II
 */
class Solution {
public:
    vector<vector<int>> generateMatrix(int n) {
        if (n < 1) return vector<vector<int>> ();
        vector<vector<int>> res {n, vector<int>(n, 0)};
        int row_begin = 0, row_end = n;
        int col_begin = 0, col_end = n;
        int begin = 1, end = n * n;
        while (begin <= end){
            for (int col = col_begin; col < col_end; col++)
                res[row_begin][col] = begin++;
            for (int row = row_begin + 1; row < row_end ; row++)
                res[row][col_end - 1] = begin++;
            for (int col = col_end - 2; col >= col_begin; col--)
                res[row_end - 1][col] = begin++;
            for (int row = row_end - 2; row > row_begin; row--)
                res[row][col_begin] = begin++;
            row_begin++, col_begin++, row_end--, col_end--; 
        }
        return res;
    }
};

File: Difficulty/Medium/[59] Spiral Matrix II/solution.py
class Solution:
    def generateMatrix(self, n: int) -> List[List[int]]:
        if n < 1:
            return []
        res = [n * [0] for _ in range(n)]
        row_begin, row_end, col_begin, col_end, i = 0, n, 0, n, 1
        while i <= n**2:
            for col in range(col_begin, col_end):
                res[row_begin][col] = i
                i += 1
            for row in range(row_begin + 1, row_end):
                res[row][col_end - 1] = i
                i += 1
            for col in range(col_end - 2, col_begin - 1, -1):
                res[row_end - 1][col] = i
                i += 1
            for row in range(row_end - 2, row_begin, -1):
                res[row][col_begin] = i
                i += 1
            row_begin += 1
            col_begin += 1
            row_end -= 1
            col_end -= 1
        return res